pkg/pki/vaultpki/vault_common.go

```diff
@@ -0,0 +1,335 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vaultpki
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/banzaicloud/kafka-operator/api/v1alpha1"
+	"github.com/banzaicloud/kafka-operator/pkg/errorfactory"
+	pkicommon "github.com/banzaicloud/kafka-operator/pkg/util/pki"
+	vaultapi "github.com/hashicorp/vault/api"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	vaultBackendPKI = "pki"
+	vaultBackendKV  = "kv"
+
+	// Fields in a vault PKI certificate
+	vaultCertificateKey = "certificate"
+	vaultIssuerKey      = "issuing_ca"
+	vaultPrivateKeyKey  = "private_key"
+	vaultSerialNoKey    = "serial_number"
+
+	// Common arguments to vault certificates/issuers/configs
+	vaultCommonNameArg        = "common_name"
+	vaultAltNamesArg          = "alt_names"
+	vaultTTLArg               = "ttl"
+	vaultOUArg                = "ou"
+	vaultPrivateKeyFormatArg  = "private_key_format"
+	vaultExcludeCNFromSANSArg = "exclude_cn_from_sans"
+
+	// Arguments to issuer config
+	vaultIssuingCertificates   = "issuing_certificates"
+	vaultCRLDistributionPoints = "crl_distribution_points"
+
+	// Arguments to create issue role
+	vaultAllowLocalhost   = "allow_localhost"
+	vaultAllowedDomains   = "allowed_domains"
+	vaultAllowSubdomains  = "allow_subdomains"
+	vaultMaxTTL           = "max_ttl"
+	vaultAllowAnyName     = "allow_any_name"
+	vaultAllowIPSANS      = "allow_ip_sans"
+	vaultAllowGlobDomains = "allow_glob_domains"
+	vaultOrganization     = "organization"
+	vaultUseCSRCommonName = "use_csr_common_name"
+	vaultUseCSRSANS       = "use_csr_sans"
+)
+
+// getCAPath returns the path to the pki mount for a cluster
+func (v *vaultPKI) getCAPath() string {
+	return fmt.Sprintf("pki_%s", v.cluster.GetUID())
+}
+
+// getUserStorePath returns the path to where issued keys/certificates are
+// held persistently for use within the operator
+func (v *vaultPKI) getUserStorePath() string {
+	return fmt.Sprintf("pki_users_%s", v.cluster.GetUID())
+}
+
+// getIssuePath returns the path for issuing certificates for the cluster
+func (v *vaultPKI) getIssuePath() string {
+	return fmt.Sprintf("%s/issue/operator", v.getCAPath())
+}
+
+// getRevokePath returns the revocation path for a cluster
+func (v *vaultPKI) getRevokePath() string {
+	return fmt.Sprintf("%s/revoke", v.getCAPath())
+}
+
+// getStorePathForUser returns the persistent storage path for a user
+func (v *vaultPKI) getStorePathForUser(user *v1alpha1.KafkaUser) string {
+	return fmt.Sprintf("%s/%s", v.getUserStorePath(), user.GetUID())
+}
+
+// checkSecretPath checks if the provided path is a full vault path and
+// returns a modified path for the default backend if not
+func checkSecretPath(path string) string {
+	if len(strings.Split(path, "/")) == 1 {
+		return fmt.Sprintf("secret/%s", path)
+	}
+	return path
+}
+
+// getClient retrieves a vault client using the environment configuration
```

Comment:
For this purpose (service account-based authentication) I would pull in the https://github.com/banzaicloud/bank-vaults/blob/master/pkg/sdk/vault/client.go#L106, it does it for you and renews the token automatically.

Reply:
Will take a look at that 😄